libraries/lib-module-manager/PluginIPCUtils.cpp

```diff
@@ -0,0 +1,157 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file PluginIPCUtils.cpp
+
+  @author Vitaly Sverchinsky
+  
+  Part of lib-module-manager library.
+
+**********************************************************************/
+
+#include "PluginIPCUtils.h"
+
+#include "PluginDescriptor.h"
+
+#include "IPCChannel.h"
+#include "XMLWriter.h"
+
+using namespace details;
+
+namespace {
+   using HeaderBlock = size_t;
+   constexpr auto HeaderBlockSize = sizeof(HeaderBlock);
+
+   constexpr auto NodePlugin = "Plugin";
+   constexpr auto NodeError = "Error";
+   constexpr auto AttrErrorMessage = "msg";
+}
+
+bool details::ParseRequestString(const wxString& req, wxString& providerId, wxString& pluginPath)
+{
+   auto strings = wxSplit(req, ';');
+   if(strings.size() == 2)
+   {
+      providerId = strings[0];
+      pluginPath = strings[1];
+      return true;
+   }
+   return false;
+}
+
+wxString details::MakeRequestString(const wxString& providerId, const wxString& pluginPath)
+{
+   return wxJoin(wxArrayStringEx {providerId, pluginPath}, ';');
+}
+
+void details::PutMessage(IPCChannel& channel, const wxString& value)
+{
+   auto utf8 = value.ToUTF8();
+   const HeaderBlock length = utf8.length();
+   channel.Send(&length, HeaderBlockSize);
+   channel.Send(utf8.data(), length);
+}
+
+void InputMessageReader::ConsumeBytes(const void* bytes, size_t length)
+{
+   auto from = mBuffer.size();
+   mBuffer.resize(from + length);
+   std::memcpy(&mBuffer[from], bytes, length);
+}
+
+bool InputMessageReader::CanPop() const noexcept
+{
+   if(mBuffer.size() >= HeaderBlockSize)
+   {
+      auto expectedSize = *reinterpret_cast<const HeaderBlock*>(mBuffer.data());
+      return mBuffer.size() >= expectedSize + HeaderBlockSize;
+   }
+   return false;
+}
+
+wxString InputMessageReader::Pop()
+{
+   wxString message;
+   const auto length = *reinterpret_cast<const HeaderBlock*>(mBuffer.data());
+   assert(mBuffer.size() >= length + HeaderBlockSize);
```

Comment:
An assertion without proof -- but sufficient "proof" for me would be that you document in the header file,

```
//! @pre `CanPop()`
```

So the burden of further proof is properly on the caller.